### DISCUSSION TITLE
Strange metadata API

### DISCUSSION THREAD

**Discussion by benoj** [saleor/saleor#9223] [General]
### What I'm trying to achieve
Reading values from the metadata on the front end

### Steps to reproduce the problem
1. Add metadata to something which can have it
2. query using graphql API

### What I expected to happen

The metadata API feels....weird. Not sure what the logic is in having this structure:

```
[
    {
        "key": "myKey",
        "value": "myValue"
    },
    ...
]
```
over a hashmap type structure:
```
{
   "myKey": "myValue", 
   ...
}
```

At present, we store a lot of pieces of information about the order and products in the metadata so we have a lot of code like this:

```
const weightGold: any = line?.variant?.metadata.find(m => m?.key === "weight.gold")?.value // O(n)
```

which is a bit of a mouthful and not necessarily that performant (although let's be honest if metadata was so big that it needed a performant search that might be a smell)

It feels as though a dictionary structure would simplify a whole lot of things:

```
const weightGold: any = line?.variant?.metadata["weight.gold"] // O(1)-ish
```

And potentially make the DB structure more simple (i.e. metadata could become a Postgres JSON column)

Just wondering if there was some other reason that it has been implemented in this manner

Thanks,

Ben

**Comment by LeOndaz** [comment]
Interesting..

**Comment by patrys** [comment]
There is no map type in GraphQL. You can, however, write a client-side function that unpacks it to a flat object in JS.

**Comment by benoj** [comment]
Seems a little odd to me to make the client side life harder when the type of Metadata by definition is untyped map data.

There exists examples of graphql json data for sure in node projects https://github.com/taion/graphql-type-json I’m presuming some will exist for python?

**Comment by patrys** [comment]
When consuming custom GraphQL scalar types (like JSON), you still need to write a parser for the values (and write custom types for TypeScript if you use it) as custom scalars are transferred as text blobs. The way it's implemented currently is the natural representation of a map in GraphQL.

**Comment by benoj** [comment]
We do use typescript - where we are using untyped map data it's supported in typescript with `{[key: string]: any}` or as native json in javascript.

I think it's safe to assume that 99% of the clients for this will UI's built-in js (or some superset of it e.g. typescript etc..) where JSON is a "native" structure. 


This feels more intuitive to me:

```javascript
const response = await fetch(
  'https://mysaleorbackend.com/graphql/',
  {
    method: 'post',
    body: query,
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': data.length,
    },
  }
);

const item = response.?whatever.?metadata['my-item']
```

over 
```javascript
const response = await fetch(
  'https://mysaleorbackend.com/graphql/',
  {
    method: 'post',
    body: query,
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': data.length,
    },
  }
);

const item = response.?whatever.?metadata.find(m => m?.key === "my-item")?.value
```

or  this (which feels unnecessary)

```javascript
const processMeta =. (meta) => meta.reduce((agg, item) => {
    agg[item.key] = item.value
    return agg
}, {})

const response = await fetch(
  'https://mysaleorbackend.com/graphql/',
  {
    method: 'post',
    body: query,
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': data.length,
    },
  }
);

const flatMetadata = processMeta(response?.whatever?.metadata)
const item = flatMetadata['my-item']

```

**Comment by patrys** [comment]
Ah, I see. If you're using raw values returned by fetch, then indeed, it looks like more work. I'll see if we can provide an alternative interface to fetch this type of data that does not rely on either manual flattening (metadata) or unmarshalling (product descriptions).
